### PR TITLE
Pretty CLOSE :P

### PR DESCRIPTION
**Pull request by raicho81**

### CODE REVIEW

**Comment by raicho81** [comment]
image

**Comment by raicho81** [comment]
image

**Comment by raicho81** [comment]
image

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# Install Nim
+RUN wget -q https://nim-lang.org/download/nim-2.0.0-linux_x64.tar.xz
+RUN ls -al
+RUN xz -d nim-2.0.0-linux_x64.tar.xz
+RUN tar -xf nim-2.0.0-linux_x64.tar
+
+# Command to run on container start
+# ENTRYPOINT[]"/app/nim-2.0.0-linux_x64/bin/nim"]
+# CMD ["--version"]
+ENTRYPOINT ["tail", "-f", "/dev/null"]
```
Comment (zah) [inline Dockerfile:40]: What's the intended use of this Dockerfile and why do we need the zip below? Some comments hear or an expansion of the project README would be helpful.

File: misc/http_client.nim [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# nim-raft
```
Comment (zah) [inline misc/http_client.nim:1]: How is this file related to the project? It look like some personal code experiment.

File: raft/consensus_module.nim [hunk 3]
```diff
@@ -5,4 +5,95 @@
...
+  result = false
+  withRLock(node.raftStateMutex):
+    var cnt = 0
+    for peer in node.peers:
+      if peer.hasVoted:
+        cnt.inc
+    if cnt >= (node.peers.len div 2 + node.peers.len mod 2):
+      result = true
+
+proc RaftNodeHandleHeartBeat*[SmCommandType, SmStateType](node: RaftNode[SmCommandType, SmStateType], msg: RaftMessageAppendEntries): RaftMessageAppendEntriesResponse[SmStateType] =
```
Comment (zah) [inline raft/consensus_module.nim:25]: According to [NEP-1](https://nim-lang.org/docs/nep1.html), function names should use `camelCase` in Nim.

File: raft/consensus_module.nim [hunk 4]
```diff
@@ -5,4 +5,95 @@
...
+# those terms.
+
+import types
+import protocol
+import log_ops
+import chronicles
+
+proc RaftNodeQuorumMin[SmCommandType, SmStateType](node: RaftNode[SmCommandType, SmStateType]): bool =
+  result = false
+  withRLock(node.raftStateMutex):
```
Comment (zah) [inline raft/consensus_module.nim:17]: Please note that the current code is not thread safe because `RaftNode` is a ref type. You cannot share the same reference between multiple threads in Nim.

You could argue perhaps that Nim 2.0 in ORC mode changes the semantics of `ref` to make this possible, but then I would say that using `withRLock` directly is not-appropriate. You can use some alternative template that compiles to a no-op if ORC/ARC is not enabled.

File: raft/consensus_module.nim [hunk 5]
```diff
@@ -5,4 +5,95 @@
...
+          RaftMessageRequestVote(
+            op: rmoRequestVote, msgId: genUUID(), senderId: node.id,
+            receiverId: peer.id, lastLogTerm: RaftNodeLogEntryGet(node, RaftNodeLogIndexGet(node)).term,
+            lastLogIndex: RaftNodeLogIndexGet(node), senderTerm: node.currentTerm)
+        )
+      )
+
+  # Process votes (if any)
+  for voteFut in node.votesFuts:
+    let r = await voteFut
```
Comment (zah) [inline raft/consensus_module.nim:75]: Is it important to process the votes in certain order (e.g. order of arrival)? The code here will process them in order of sending and if the first vote doesn't arrive, none of the votes will be processed. Also, could we wait here forever? Is there a timeout being applied anywhere?

File: raft/raft_api.nim [hunk 6]
```diff
@@ -9,62 +9,160 @@
...
-proc RaftNodeCreateNew*[LogEntryDataType, SmStateType](                     # Create New Raft Node
-                  id: RaftNodeId, peers: RaftNodePeers,
-                  persistentStorage: RaftNodePersistentStorage,
-                  msgSendCallback: RaftMessageSendCallback): RaftNode[LogEntryDataType, SmStateType] =
-  discard
-
-proc RaftNodeLoad*[LogEntryDataType, SmStateType](
+import consensus_module
+import log_ops
+import ../db/kvstore_mdbx
```
Comment (zah) [inline raft/raft_api.nim:14]: I think the persistence module should be a bit more optional. For example, the base library can define some kind of callback that's responsible for implementing the persistence and then the user can import a helper module implementing the callback with a particular database engine.

File: tests/tornado_simple_raft_node_server.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import asyncio
```
Comment (zah) [inline tests/tornado_simple_raft_node_server.py:1]: Some comments here or in the README would be good to explain the intended usage of this server.

File: tests/test_basic_cluster.nim [hunk 8]
```diff
@@ -0,0 +1,41 @@
+# nim-raft
+# Copyright (c) 2023 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import unittest2
```
Comment (zah) [inline tests/test_basic_cluster.nim:10]: These tests looks like simulations and as actual tests. No acceptance criteria is defined (i.e. what the conditions that would make the test fail?)

**Comment by zah** [comment]
Many of the existing Raft libraries split the state machine code from the networking code, so it can be tested in isolation: 
https://github.com/scylladb/scylladb/blob/acf8da2bcef47168376e876e2fdc37fc26ebc82e/raft/fsm.hh

This FSM layer probably doesn't need to have any async procedures as well. The time can also be supplied by the client code in order to allow the tests to speed up the passage of time arbitrarily:

Here is a relevant excerpt from the file above:

```
// Seastar Raft instead implements an instance of Raft as
// in-memory state machine with a catch-all API step(message)
// method. The method handles any kind of input and performs the
// needed state machine state transitions. To get state machine output
// poll_output() function has to be called. This call produces an output
// object, which encapsulates a list of actions that must be
// performed until the next poll_output() call can be made. The time is
// represented with a logical timer. The client is responsible for
// periodically invoking tick() method, which advances the state
// machine time and allows it to track such events as election or
// heartbeat timeouts.
```
I don't know if you payed attention but I have just an interface to the SM and it is currently split from the other code and can be tested in isolation and I have a simple SM implementation including a test for it. Nevertheless I will check this out because it looks more advanced than my current idea.

File: Dockerfile (hunk 1 above)
Comment (raicho81) [reply Dockerfile:40]: Sorry zip is redundant will remove it.

Comment (raicho81) [reply Dockerfile:40]: I was planning to run the Raft Nodes in cluster that's why I've put the Dockerfile. It is not done yet.

File: misc/http_client.nim (hunk 2 above)
Comment (raicho81) [reply misc/http_client.nim:1]: Right I will remove it.

File: raft/consensus_module.nim (hunk 3 above)
Comment (raicho81) [reply raft/consensus_module.nim:25]: Ok will fix it.

File: raft/consensus_module.nim (hunk 4 above)
Comment (raicho81) [reply raft/consensus_module.nim:17]: Ok I will consider this notes and try to make it like this.

File: raft/consensus_module.nim (hunk 5 above)
Comment (raicho81) [reply raft/consensus_module.nim:75]: Yes you are right I was thinking the same - that I've forgotten to put Timeout for the response. We can't wait for ever this is true.

File: raft/raft_api.nim (hunk 6 above)
Comment (raicho81) [reply raft/raft_api.nim:14]: Ok I will change it np.

File: tests/test_basic_cluster.nim (hunk 8 above)
Comment (raicho81) [reply tests/test_basic_cluster.nim:10]: I am not done with this test yet.
